Expose fork method on python API and sync windows state with backend

ricardofbarros opened the pull request:
## Description
This PR will add the following features: 
- expose a `.fork()` method on the client python API with the same functionality as the UI.
- sync layout state with the backend. We can resize the windows and reload the page that the windows will keep the layout state.

## Motivation and Context
#323

## How Has This Been Tested?
Installing visdom directly from my repo and testing by hand.

## Screenshots (if appropriate):

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] For JavaScript changes, I have re-generated the minified JavaScript code.

ricardofbarros commented:
cc @JackUrb

ricardofbarros commented:
I just need some guidance on what should the server reply back to the client on the following cases:

- https://github.com/facebookresearch/visdom/pull/609/files#diff-ddbe57b92511a8307bcf5999eb519e6cR860
- https://github.com/facebookresearch/visdom/pull/609/files#diff-ddbe57b92511a8307bcf5999eb519e6cR863

JackUrb commented on README.md:
```diff
@@ -631,6 +631,14 @@ This function deletes a specified env entirely. It takes env id `eid` as input.
 > **Note**: `delete_env` is deletes all data for an environment and is IRREVERSIBLE. Do not use unless you absolutely want to remove an environment.
 
 
+#### vis.fork
```
I think it would be more appropriate to use `fork_env` as we may have something specifically for forking layouts in the future. CamelCase locations should then use `ForkEnv`.

JackUrb commented on py/visdom/server.py:
```diff
@@ -840,6 +845,29 @@ def post(self):
+        eid = args.get('eid')
+
+        if prev_eid not in handler.state:
+            return handler.write('TODO: env doesn\'t exit')
```
You can use an assert in this case, the error should propagate to the visdom client, which I think is an appropriate outcome to trying to fork an environment that doesn't exist. Note in the docs that such an exception will occur if an env that doesn't exist is forked.

JackUrb commented on py/visdom/server.py:
```diff
@@ -840,6 +845,29 @@ def post(self):
+            return handler.write('TODO: env doesn\'t exit')
+
+        handler.state[eid] = copy.deepcopy(handler.state[prev_eid])
+        handler.write('TODO: missing proper reply')
```
Proper reply here is probably either `True` or the new env_id. As the failure case will be an exception, there's no real requirement for the success case.

JackUrb commented on py/visdom/server.py:
```diff
@@ -856,11 +856,10 @@ def wrap_func(handler, args):
-            return handler.write('TODO: env doesn\'t exit')
+        assert prev_eid in handler.state, 'env to be forked doesn\'t exit'
 
         handler.state[eid] = copy.deepcopy(handler.state[prev_eid])
```
Just realizing you're missing a few small bits based on the implementation of fork that's present in SocketHandler:
```
eid = escape_eid(eid)
```
is used to make sure that the eid is valid before it's put in.
```
serialize_env(self.state, [eid], env_path=self.app.env_path)
```
ensures the new env is written to file once it's created, and 
```
broadcast_envs(self)
```
should ensure that open windows will receive the forked environment.

ricardofbarros replied:
Fixed :)

ricardofbarros commented:
@JackUrb I thank you for guiding me through this feature :) 

I've fixed what you requested and I also noticed I coded a bug on handling `layout_item_update` on backend. 

It was overwriting the layout state of the env with a single window. I fixed that as well and now the feature works perfectly in any scenario of a fork. 

Before that fix if I resized a window and if I immediately forked that env, without saving, the new env wouldn't pick up those changes.

ricardofbarros commented:
Any ETA for this @JackUrb ?

JackUrb commented:
Sorry for the delay @ricardofbarros - I've been split across a number of different tasks. I'll get to it this week in my conference downtime

ricardofbarros commented:
Thanks for the reply :)
